File: js/LinkedList/146.lru-cache.ts
class node<T> {
  key: T
  val: T
  prev: node<T> = null
  next: node<T> = null
  constructor(key: T, value: T) {
    this.key = key
    this.val = value
  }
}
class LRUCache {
  left = new node(0, 0)
  right = new node(0, 0)
  cache = new Map<number, node<number>>()
  cap = 0
  constructor(capacity: number) {
    this.cap = capacity
    this.left.next = this.right
    this.right.prev = this.left
  }
  get(key: number) {
    if (!this.cache.has(key)) {
      return -1
    }
    let node = this.cache.get(key)
    this.remove(node)
    this.insert(node)
    return node.val
  }
  put(key: number, value: number) {
    if (this.cache.has(key)) {
      this.remove(this.cache.get(key))
    }
    let n = new node(key, value)
    this.cache.set(key, n)
    this.insert(n)
    if (this.cache.size > this.cap) {
      let lru = this.left.next
      this.remove(lru)
      this.cache.delete(lru.key)
    }
  }
  remove(n: node<number>) {
    let { prev, next } = n
    prev.next = next
    next.prev = prev
  }
  insert(n: node<number>) {
    let [prev, next] = [this.right.prev, this.right]
    prev.next = n
    next.prev = n
    n.prev = prev
    n.next = next
  }
}

File: cpp/LinkedList/146.lru-cache.cpp
#include "../std.hpp"
struct Node {
    int k;
    int val;
    Node *prev = nullptr;
    Node *next = nullptr;
    Node(int key, int value) : k(key), val(value) {}
};

class LRUCache {
public:
    LRUCache(int capacity) : cap(capacity)
    {
        ios::sync_with_stdio(false);
        cin.tie(nullptr);
        cout.tie(nullptr);
        left->next = right;
        right->prev = left;
    }

    int get(int key)
    {
        if (cache.find(key) != cache.end()) {
            auto node = cache[key];
            remove(node);
            insert(node);
            return node->val;
        }
        return -1;
    }

    void put(int key, int value)
    {
        if (cache.find(key) != cache.end()) {
            remove(cache[key]);
        }
        cache[key] = new Node(key, value);
        insert(cache[key]);
        if (cache.size() > cap) {
            // remove from list & delete LRU from map
            Node *lru = left->next;
            remove(lru);
            cache.erase(lru->k);
            delete lru;
        }
    }

private:
    int cap;
    unordered_map<int, Node *> cache; // {key -> node}
    Node *left = new Node(0, 0);
    Node *right = new Node(0, 0);
    void remove(Node *node)
    {
        Node *prev = node->prev;
        Node *next = node->next;
        prev->next = next;
        next->prev = prev;
    }
    void insert(Node *node)
    {
        Node *prev = right->prev;
        Node *next = right;
        prev->next = node;
        next->prev = node;
        node->prev = prev;
        node->next = next;
    }
};